### PR TITLE
Moving WSAGetLastError calls closer to socket operation

### PR DESCRIPTION
**Pull request by henricm**
* When doing async connect on windows, connect returns `WSAEINPROGRESS`
which needs to be returned to managed code for correct handling. However
on Windows, last error gets reset before `WSAGetLastError` is called.
This fix moves all `WSAGetLastError` calls up closer to the socket
operation to make sure correct error code is returned to the managed code.

* Making sure `TcpNoDelay` is set to false for IPv4 by default and
that `DualMode` is default for IPv6.

* Making sure option value is set to 0 before calling getsockopt
because on Windows, a `BOOL` option value (eg `TcpNoDelay`) is returned
as a one byte value and the other bytes are left untouched which resulted
in `TcpNoDelay` always seemed to be `true`.

* Fixed so `TF_REUSE_SOCKET` is set only if `reuse` parameter is set
when disconnecting a socket.

### CODE REVIEW

**Comment by marek-safar** [comment]
@esdrubal could you review the fix

**Comment by henricm** [comment]
Any update of the review of this?

File: mono/metadata/socket-io.c [hunk 1]
```diff
@@ -1423,10 +1420,10 @@ ves_icall_System_Net_Sockets_Socket_Disconnect_internal (SOCKET sock, MonoBoolea
 	MONO_ENTER_GC_SAFE;
 
 	if (_wapi_disconnectex != NULL) {
-		if (!_wapi_disconnectex (sock, NULL, TF_REUSE_SOCKET, 0))
+		if (!_wapi_disconnectex (sock, NULL, reuse ? TF_REUSE_SOCKET : 0, 0))
```
Comment (ludovic-henry) [inline mono/metadata/socket-io.c]: does this has to do with moving WSAGetLastError, or is it another fix? If it's another fix, please split into a different commit (same PR is fine)

**Comment by ludovic-henry** [comment]
LGTM
